app/code/Magento/GiftMessage/Test/Unit/Observer/SalesEventOrderItemToQuoteItemObserverTest.php

```diff
@@ -0,0 +1,225 @@
+use Magento\GiftMessage\Model\MessageFactory;
+use Magento\GiftMessage\Observer\SalesEventOrderItemToQuoteItemObserver;
+use Magento\Quote\Model\Quote\Item as QuoteItem;
+use Magento\Sales\Model\Order;
+use Magento\Sales\Model\Order\Item as OrderItem;
+use Magento\Store\Model\Store;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ */
+class SalesEventOrderItemToQuoteItemObserverTest extends TestCase
+{
+    /**
+     * Stub message id
+     */
+    private const STUB_MESSAGE_ID = 1;
+
+    /**
+     * Stub new message id
+     */
+    private const STUB_NEW_MESSAGE_ID = 2;
+
+    /**
+     * @var SalesEventOrderItemToQuoteItemObserver
+     */
+    private $observer;
+
+    /**
+     * @var MessageFactory|MockObject
+     */
+    private $messageFactoryMock;
+
+    /**
+     * @var MessageHelper|MockObject
+     */
+    private $giftMessageHelperMock;
+
+    /**
+     * @var Observer|MockObject
+     */
+    private $observerMock;
+
+    /**
+     * @var Event|MockObject
+     */
+    private $eventMock;
+
+    /**
+     * @var Order|MockObject
+     */
+    private $orderMock;
+
+    /**
+     * @var OrderItem|MockObject
+     */
+    private $orderItemMock;
+
+    /**
+     * @var Store|MockObject
+     */
+    private $storeMock;
+
+    /**
+     * @var MessageInterface|MockObject
+     */
+    private $messageMock;
+
+    /**
+     * @var QuoteItem|MockObject
+     */
+    private $quoteItemMock;
+
+    /**
+     * Prepare environment for test
+     */
+    public function setUp(): void
+    {
+        $this->messageFactoryMock = $this->getMockBuilder(MessageFactory::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['create'])
+            ->getMock();
+        $this->giftMessageHelperMock = $this->createMock(MessageHelper::class);
+        $this->observerMock = $this->createMock(Observer::class);
+        $this->eventMock = $this->createPartialMock(Event::class, ['getOrderItem', 'getQuoteItem']);
+        $this->orderItemMock = $this->createPartialMock(
+            OrderItem::class,
+            ['getOrder', 'getStoreId', 'getGiftMessageId']
+        );
+        $this->quoteItemMock = $this->createPartialMock(QuoteItem::class, ['setGiftMessageId']);
+        $this->orderMock = $this->createPartialMock(Order::class, ['getReordered']);
+        $this->storeMock = $this->createMock(Store::class);
+        $this->messageMock = $this->createMock(MessageModel::class);
+
+        $this->eventMock->expects($this->atLeastOnce())
+            ->method('getOrderItem')
+            ->willReturn($this->orderItemMock);
+
+        $this->orderItemMock->expects($this->atLeastOnce())
+            ->method('getOrder')
+            ->willReturn($this->orderMock);
+
+        $this->observerMock->expects($this->atLeastOnce())
+            ->method('getEvent')
+            ->willReturn($this->eventMock);
```

Comment:
For these three, I'd avoid `->expects` part, as someone could want to test the scenario where code does not reach the `getOrderItem` or `getOrder` point.